pygments/lexers/ul4.py

```diff
@@ -0,0 +1,261 @@
+        "ul4": [
+            # End the tag
+            (r"\?>", token.Comment.Preproc, "#pop"),
+            # Start triple quoted string constant
+            ("'''", token.String, "string13"),
+            ('""""', token.String, "string23"),
+            # Start single quoted string constant
+            ("'", token.String, "string1"),
+            ('"', token.String, "string2"),
+            # Floating point number
+            (r"\d+\.\d*([eE][+-]?\d+)?", token.Number.Float),
+            (r"\.\d+([eE][+-]?\d+)?", token.Number.Float),
+            (r"\d+[eE][+-]?\d+", token.Number.Float),
+            # Binary integer: ``0b101010``
+            (r"0[bB][01]+", token.Number.Bin),
+            # Octal integer: ``0o52``
+            (r"0[oO][0-7]+", token.Number.Oct),
+            # Hexadecimal integer: ``0x2a``
+            (r"0[xX][0-9a-fA-F]+", token.Number.Hex),
+            # Date or datetime: ``@(2000-02-29)``/``@(2000-02-29T12:34:56.987654)``
+            (r"@\(\d\d\d\d-\d\d-\d\d(T(\d\d:\d\d(:\d\d(\.\d{6})?)?)?)?\)", token.Literal.Date),
+            # Color: ``#fff``, ``#fff8f0`` etc.
+            (r"#[0-9a-fA-F]{8}", token.Literal.Color),
+            (r"#[0-9a-fA-F]{6}", token.Literal.Color),
+            (r"#[0-9a-fA-F]{3,4}", token.Literal.Color),
+            # Decimal integer: ``42``
+            (r"\d+", token.Number.Integer),
+            # Operators
+            (r"//|==|!=|>=|<=|<<|>>|\+=|-=|\*=|/=|//=|<<=|>>=|&=|\|=|^=|=|[\[\]{},:*/().~%&|<>^+-]", token.Token.Operator),
+            # Keywords
+            (lexer.words(("for", "in", "if", "else", "not", "is", "and", "or"), suffix=r"\b"), token.Keyword),
+            # Builtin constants
+            (lexer.words(("None", "False", "True"), suffix=r"\b"), token.Keyword.Constant),
+            # Variable names
+            (r"[a-zA-Z_][a-zA-Z0-9_]*", token.Name),
+            # Whitespace
+            (r"\s+", token.Text),
+        ],
+        # ``<?end ...?>`` tag for closing the last open block
+        "end": [
+            (r"\?>", token.Comment.Preproc, "#pop"),
+            (lexer.words(("for", "if", "def", "while", "renderblock", "renderblocks"), suffix=r"\b"), token.Keyword),
+            (r"\s+", token.Text),
+            (r".", token.Error), # Unrecognized tag name
+        ],
+        # Content of the ``<?whitespace ...?>`` tag:
+        # ``keep``, ``string`` or ``smart``
+        "whitespace": [
+            (r"\?>", token.Comment.Preproc, "#pop"),
+            (lexer.words(("keep", "strip", "smart"), suffix=r"\b"), token.Comment.Preproc),
+            (r"\s+", token.Text),
+            (r".", token.Error), # Unrecognized whitespace mode
+        ],
+        # Inside a string constant
+        "string": [
+            ("\\\\['\"abtnfr]", token.String.Escape),
+            (r"\\x[0-9a-fA-F]{2}", token.String.Escape),
+            (r"\\u[0-9a-fA-F]{4}", token.String.Escape),
+            (r"\\U[0-9a-fA-F]{8}", token.String.Escape),
+            (r".", token.String),
```

Comment:
This results in one token per character on most strings, which can increase output size and make raw token output (in tests) hard to read. The typical fix is to use something like `r"""[^\\'"]+"""`.

Reply:
OK, I fixed this in https://github.com/doerwalter/pygments/commit/34927de02bff7b1718c34b2d3490e4e846c86d10